<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Brian J. Knaus" />

<meta name="date" content="2017-02-13" />

<title>Determining ploidy 2</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link rel="stylesheet" href="/home/local/USDA-ARS/knausb/R/x86_64-pc-linux-gnu-library/3.3/rmarkdown/rmarkdown/templates/html_vignette/resources/vignette.css" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Determining ploidy 2</h1>
<h4 class="author"><em>Brian J. Knaus</em></h4>
<h4 class="date"><em>2017-02-13</em></h4>



<p>Previously we showed how to create histograms based on the ratio of alleles observed at heterozygous positions. These histograms may present desireable perspectives when a constant ploidy is expected. Investigators interested in whether there is variation in copy number along a chromosome may require a more detailed perspective. Here we present a perspective of copy number variation along the length of a chromosome.</p>
<p>Instead of using all of the data each chromosome can be divided into windows of user specified width. For each of these windows a numerical summary of where the peak would be is made. Creating a numerical summary is important because when we move to genome scale data it is easy to generate more windows than we can manually curate. We can visualize some of this data, but the numerical summary will allow us to process many windows over many samples.</p>
<div id="input-data" class="section level2">
<h2>Input data</h2>
<p>Data import is performed similar to other examples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load libraries</span>
<span class="kw">library</span>(vcfR)
<span class="kw">library</span>(pinfsc50)

<span class="co"># Determine file locations</span>
vcf_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;pinf_sc50.vcf.gz&quot;</span>,
                        <span class="dt">package =</span> <span class="st">&quot;pinfsc50&quot;</span>)

<span class="co"># Read data into memory</span>
vcf &lt;-<span class="st"> </span><span class="kw">read.vcfR</span>(vcf_file, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
vcf</code></pre></div>
<pre><code>## ***** Object of Class vcfR *****
## 18 samples
## 22,031 variants
## Object size: 20.9 Mb
## 7.929 percent missing data
## *****        *****         *****</code></pre>
</div>
<div id="depth-filtering" class="section level2">
<h2>Depth filtering</h2>
<p>Weâ€™ll also filter on depth by removng high and low coverage variants. The thresholds for this filtering will be determined by quantiles. Researchers may want to explore other methods available in R (e.g., fit a distribution, mixture models, etc.).</p>
<p>First we extract the alleles and create frequencies</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ad &lt;-<span class="st"> </span><span class="kw">extract.gt</span>(vcf, <span class="dt">element =</span> <span class="st">&#39;AD&#39;</span>)

allele1 &lt;-<span class="st"> </span><span class="kw">masplit</span>(ad, <span class="dt">record =</span> <span class="dv">1</span>)
allele2 &lt;-<span class="st"> </span><span class="kw">masplit</span>(ad, <span class="dt">record =</span> <span class="dv">2</span>)

ad1 &lt;-<span class="st"> </span>allele1 /<span class="st"> </span>(allele1 +<span class="st"> </span>allele2)
ad2 &lt;-<span class="st"> </span>allele2 /<span class="st"> </span>(allele1 +<span class="st"> </span>allele2)</code></pre></div>
<p>We create thresholds based on the most abundant allele as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dp &lt;-<span class="st"> </span>allele1
<span class="co">#sums &lt;- apply(dp, MARGIN=2, quantile, probs=c(0.15, 0.95), na.rm=TRUE)</span>
sums &lt;-<span class="st"> </span><span class="kw">apply</span>(dp, <span class="dt">MARGIN=</span><span class="dv">2</span>, quantile, <span class="dt">probs=</span><span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>), <span class="dt">na.rm=</span><span class="ot">TRUE</span>)

<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>))

for(i in <span class="dv">1</span>:<span class="dv">12</span>){
  <span class="kw">hist</span>(allele1[,i], <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="fl">1e3</span>,<span class="dt">by=</span><span class="dv">1</span>), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">100</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;&quot;</span>)
  <span class="kw">title</span>(<span class="dt">main =</span> <span class="kw">colnames</span>(allele1)[i])
  <span class="kw">abline</span>(<span class="dt">v=</span>sums[,i], <span class="dt">col=</span><span class="dv">2</span>)
}
<span class="kw">title</span>(<span class="dt">xlab =</span> <span class="st">&quot;Depth&quot;</span>, <span class="dt">line=</span><span class="dv">0</span>, <span class="dt">outer =</span> <span class="ot">TRUE</span>, <span class="dt">font=</span><span class="dv">2</span>)
<span class="kw">title</span>(<span class="dt">ylab =</span> <span class="st">&quot;Count&quot;</span>, <span class="dt">line=</span><span class="dv">0</span>, <span class="dt">outer =</span> <span class="ot">TRUE</span>, <span class="dt">font=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-3-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</code></pre></div>
<p>We performa similar operation for the second most abundant allele.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>))

for(i in <span class="dv">1</span>:<span class="dv">12</span>){
  tmp &lt;-<span class="st"> </span>allele2[,i]
  tmp &lt;-<span class="st"> </span>tmp[ tmp &gt;<span class="st"> </span><span class="dv">0</span> ]
  <span class="kw">hist</span>(tmp, <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="fl">1e3</span>,<span class="dt">by=</span><span class="dv">1</span>), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">100</span>), 
<span class="co">#       ylim=c(0,1000),</span>
       <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;&quot;</span>)
  <span class="kw">title</span>(<span class="dt">main =</span> <span class="kw">colnames</span>(allele1)[i])
}
<span class="kw">title</span>(<span class="dt">xlab =</span> <span class="st">&quot;Depth&quot;</span>, <span class="dt">line=</span><span class="dv">0</span>, <span class="dt">outer =</span> <span class="ot">TRUE</span>, <span class="dt">font=</span><span class="dv">2</span>)
<span class="kw">title</span>(<span class="dt">ylab =</span> <span class="st">&quot;Count&quot;</span>, <span class="dt">line=</span><span class="dv">0</span>, <span class="dt">outer =</span> <span class="ot">TRUE</span>, <span class="dt">font=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-4-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</code></pre></div>
</div>
<div id="find-peaks-of-density" class="section level2">
<h2>Find peaks of density</h2>
<p>In order to avoid assumptions about the distribution of the allele balance ratios we employ a non-parametric method. The data are binned into bins of user specified widths and the bin with the highest density is chosen as the frequency. This is our numerical summary of our peak density and will be retained in a matrix of samples and windows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Filter on depth quantiles.</span>
sums &lt;-<span class="st"> </span><span class="kw">apply</span>(allele1, <span class="dt">MARGIN=</span><span class="dv">2</span>, quantile, <span class="dt">probs=</span><span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>), <span class="dt">na.rm=</span><span class="ot">TRUE</span>)
<span class="co"># Allele 1</span>
dp2 &lt;-<span class="st"> </span><span class="kw">sweep</span>(allele1, <span class="dt">MARGIN=</span><span class="dv">2</span>, <span class="dt">FUN =</span> <span class="st">&quot;-&quot;</span>, sums[<span class="dv">1</span>,])
<span class="co">#allele1[dp2 &lt; 0] &lt;- NA</span>
vcf@gt[,-<span class="dv">1</span>][ dp2 &lt;<span class="st"> </span><span class="dv">0</span> &amp;<span class="st"> </span>!<span class="kw">is.na</span>(vcf@gt[,-<span class="dv">1</span>]) ] &lt;-<span class="st"> </span><span class="ot">NA</span>
dp2 &lt;-<span class="st"> </span><span class="kw">sweep</span>(allele1, <span class="dt">MARGIN=</span><span class="dv">2</span>, <span class="dt">FUN =</span> <span class="st">&quot;-&quot;</span>, sums[<span class="dv">2</span>,])
<span class="co">#allele1[dp2 &gt; 0] &lt;- NA</span>
vcf@gt[,-<span class="dv">1</span>][dp2 &gt;<span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> </span><span class="ot">NA</span>
<span class="co"># Allele 2</span>
dp2 &lt;-<span class="st"> </span><span class="kw">sweep</span>(allele2, <span class="dt">MARGIN=</span><span class="dv">2</span>, <span class="dt">FUN =</span> <span class="st">&quot;-&quot;</span>, sums[<span class="dv">1</span>,])
vcf@gt[,-<span class="dv">1</span>][ dp2 &lt;<span class="st"> </span><span class="dv">0</span> &amp;<span class="st"> </span>!<span class="kw">is.na</span>(vcf@gt[,-<span class="dv">1</span>]) ] &lt;-<span class="st"> </span><span class="ot">NA</span>
dp2 &lt;-<span class="st"> </span><span class="kw">sweep</span>(allele2, <span class="dt">MARGIN=</span><span class="dv">2</span>, <span class="dt">FUN =</span> <span class="st">&quot;-&quot;</span>, sums[<span class="dv">2</span>,])
vcf@gt[,-<span class="dv">1</span>][dp2 &gt;<span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> </span><span class="ot">NA</span>

<span class="co"># Censor homozygotes.</span>
gt &lt;-<span class="st"> </span><span class="kw">extract.gt</span>(vcf, <span class="dt">element =</span> <span class="st">&#39;GT&#39;</span>)
hets &lt;-<span class="st"> </span><span class="kw">is_het</span>(gt)
<span class="kw">is.na</span>( vcf@gt[,-<span class="dv">1</span>][ !hets ] ) &lt;-<span class="st"> </span><span class="ot">TRUE</span>




<span class="co"># Extract allele depths</span>
ad &lt;-<span class="st"> </span><span class="kw">extract.gt</span>(vcf, <span class="dt">element =</span> <span class="st">&#39;AD&#39;</span>)
allele1 &lt;-<span class="st"> </span><span class="kw">masplit</span>(ad, <span class="dt">record =</span> <span class="dv">1</span>)
allele2 &lt;-<span class="st"> </span><span class="kw">masplit</span>(ad, <span class="dt">record =</span> <span class="dv">2</span>)
ad1 &lt;-<span class="st"> </span>allele1 /<span class="st"> </span>(allele1 +<span class="st"> </span>allele2)
ad2 &lt;-<span class="st"> </span>allele2 /<span class="st"> </span>(allele1 +<span class="st"> </span>allele2)

<span class="co"># Parameters</span>
<span class="co">#winsize &lt;- 1e5</span>
<span class="co">#</span>
winsize &lt;-<span class="st"> </span><span class="fl">2e5</span>
<span class="co">#bin_width &lt;- 0.1</span>
<span class="co">#bin_width &lt;- 0.05</span>
<span class="co">#bin_width &lt;- 0.025</span>
<span class="co">#</span>
bin_width &lt;-<span class="st"> </span><span class="fl">0.02</span>
<span class="co">#bin_width &lt;- 0.01</span>


<span class="co"># Find peaks</span>
freq1 &lt;-<span class="st"> </span>ad1/(ad1+ad2)
freq2 &lt;-<span class="st"> </span>ad2/(ad1+ad2)
myPeaks1 &lt;-<span class="st"> </span><span class="kw">freq_peak</span>(freq1, <span class="kw">getPOS</span>(vcf), <span class="dt">winsize =</span> winsize, <span class="dt">bin_width =</span> bin_width)
<span class="co">#myCounts1 &lt;- freq_peak(freq1, getPOS(vcf), winsize = winsize, bin_width = bin_width, count = TRUE)</span>
<span class="kw">is.na</span>(myPeaks1$peaks[myPeaks1$counts &lt;<span class="st"> </span><span class="dv">20</span>]) &lt;-<span class="st"> </span><span class="ot">TRUE</span>
myPeaks2 &lt;-<span class="st"> </span><span class="kw">freq_peak</span>(freq2, <span class="kw">getPOS</span>(vcf), <span class="dt">winsize =</span> winsize, <span class="dt">bin_width =</span> bin_width, <span class="dt">lhs =</span> <span class="ot">FALSE</span>)
<span class="co">#myCounts2 &lt;- freq_peak(freq2, getPOS(vcf), winsize = winsize, bin_width = bin_width, count = TRUE)</span>
<span class="kw">is.na</span>(myPeaks2$peaks[myPeaks2$counts &lt;<span class="st"> </span><span class="dv">20</span>]) &lt;-<span class="st"> </span><span class="ot">TRUE</span></code></pre></div>
<p>As a proof of concep we can visualize plots of the major allele with a red horizontal line at our predicted peak. We repeat this for the minor allele.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>))

mySample &lt;-<span class="st"> &quot;BL2009P4_us23&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq1[ myPeaks1$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ],
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks1$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}

mySample &lt;-<span class="st"> &quot;DDR7602&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq1[ myPeaks1$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ], 
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks1$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}

mySample &lt;-<span class="st"> &quot;IN2009T1_us22&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq1[ myPeaks1$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ],
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks1$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}

mySample &lt;-<span class="st"> &quot;P17777us22&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq1[ myPeaks1$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ],
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks1$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}</code></pre></div>
<p><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-5-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</code></pre></div>
<p>Second most abundant allele.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>))

mySample &lt;-<span class="st"> &quot;BL2009P4_us23&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq2[ myPeaks1$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ],
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks2$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}

mySample &lt;-<span class="st"> &quot;DDR7602&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq2[ myPeaks1$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ],
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks2$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}

mySample &lt;-<span class="st"> &quot;IN2009T1_us22&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq2[ myPeaks1$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ],
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks2$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}

mySample &lt;-<span class="st"> &quot;P17777us22&quot;</span>
for(i in <span class="dv">1</span>:<span class="dv">4</span>){
  <span class="kw">hist</span>(freq2[ myPeaks2$wins[i,<span class="st">&#39;START_row&#39;</span>]:myPeaks1$wins[i,<span class="st">&#39;END_row&#39;</span>], mySample ],
       <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">col=</span><span class="dv">8</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)
  <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>),
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
  <span class="kw">abline</span>(<span class="dt">v=</span>myPeaks2$peaks[i,mySample], <span class="dt">col=</span><span class="dv">2</span>)
  if(i==<span class="dv">2</span>){ <span class="kw">title</span>(<span class="dt">main=</span>mySample) }
}</code></pre></div>
<p><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-6-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</code></pre></div>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>There are a number of ways to visualize this data. One way is to turn our histogram on its side, as a form of marginal summary, and plot our window summaries.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">i &lt;-<span class="st"> </span><span class="dv">2</span>

<span class="kw">layout</span>(<span class="kw">matrix</span>(<span class="dv">1</span>:<span class="dv">2</span>, <span class="dt">nrow=</span><span class="dv">1</span>), <span class="dt">widths =</span> <span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">0</span>))

mySample &lt;-<span class="st"> </span><span class="kw">colnames</span>(freq1)[i]
<span class="kw">plot</span>(<span class="kw">getPOS</span>(vcf), freq1[,mySample], <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">type=</span><span class="st">&quot;n&quot;</span>, <span class="dt">yaxt=</span><span class="st">&#39;n&#39;</span>, 
     <span class="dt">main =</span> mySample, <span class="dt">xlab =</span> <span class="st">&quot;POS&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Allele balance&quot;</span>)
<span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>), 
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">c</span>(<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>), <span class="dt">col=</span><span class="dv">8</span>)
<span class="kw">points</span>(<span class="kw">getPOS</span>(vcf), freq1[,mySample], <span class="dt">pch =</span> <span class="dv">20</span>, <span class="dt">col=</span> <span class="st">&quot;#A6CEE344&quot;</span>)
<span class="kw">points</span>(<span class="kw">getPOS</span>(vcf), freq2[,mySample], <span class="dt">pch =</span> <span class="dv">20</span>, <span class="dt">col=</span> <span class="st">&quot;#1F78B444&quot;</span>)
<span class="kw">segments</span>(<span class="dt">x0=</span>myPeaks1$wins[,<span class="st">&#39;START_pos&#39;</span>], <span class="dt">y0=</span>myPeaks1$peaks[,mySample],
         <span class="dt">x1=</span>myPeaks1$wins[,<span class="st">&#39;END_pos&#39;</span>], <span class="dt">lwd=</span><span class="dv">3</span>)
<span class="kw">segments</span>(<span class="dt">x0=</span>myPeaks1$wins[,<span class="st">&#39;START_pos&#39;</span>], <span class="dt">y0=</span>myPeaks2$peaks[,mySample],
         <span class="dt">x1=</span>myPeaks1$wins[,<span class="st">&#39;END_pos&#39;</span>], <span class="dt">lwd=</span><span class="dv">3</span>)

bp1 &lt;-<span class="st"> </span><span class="kw">hist</span>(freq1[,mySample], <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">plot =</span> <span class="ot">FALSE</span>)
bp2 &lt;-<span class="st"> </span><span class="kw">hist</span>(freq2[,mySample], <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">by=</span>bin_width), <span class="dt">plot =</span> <span class="ot">FALSE</span>)

<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">2</span>))
<span class="kw">barplot</span>(<span class="dt">height=</span>bp1$counts, <span class="dt">width=</span><span class="fl">0.02</span>,  <span class="dt">space =</span> <span class="dv">0</span>, <span class="dt">horiz =</span> T, <span class="dt">add =</span> <span class="ot">FALSE</span>, <span class="dt">col=</span><span class="st">&quot;#A6CEE3&quot;</span>)
<span class="kw">barplot</span>(<span class="dt">height=</span>bp2$counts, <span class="dt">width=</span><span class="fl">0.02</span>,  <span class="dt">space =</span> <span class="dv">0</span>, <span class="dt">horiz =</span> T, <span class="dt">add =</span> <span class="ot">TRUE</span>, <span class="dt">col=</span><span class="st">&quot;#1F78B4&quot;</span>)</code></pre></div>
<p><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-7-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<p>Once we determine how to make one plot we can loop over all the samples.</p>
<p><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-1.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-2.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-3.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-4.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-5.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-6.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-7.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-8.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-9.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-10.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-11.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-12.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-13.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-14.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-15.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-16.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-17.png" style="display: block; margin: auto;" /><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-8-18.png" style="display: block; margin: auto;" /></p>
<p>We can also create a matrix of plots.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>))


for(i in <span class="dv">1</span>:<span class="dv">12</span>){
mySample &lt;-<span class="st"> </span><span class="kw">colnames</span>(freq1)[i]
<span class="kw">plot</span>(<span class="kw">getPOS</span>(vcf), freq1[,mySample], <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">type=</span><span class="st">&quot;n&quot;</span>, <span class="dt">yaxt=</span><span class="st">&#39;n&#39;</span>, 
     <span class="dt">main =</span> mySample, <span class="dt">xlab =</span> <span class="st">&quot;POS&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Allele balance&quot;</span>)
<span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">at=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>,<span class="dv">1</span>), 
     <span class="dt">labels=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="st">&#39;1/4&#39;</span>,<span class="st">&#39;1/3&#39;</span>,<span class="st">&#39;1/2&#39;</span>,<span class="st">&#39;2/3&#39;</span>,<span class="st">&#39;3/4&#39;</span>,<span class="dv">1</span>), <span class="dt">las=</span><span class="dv">1</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">c</span>(<span class="fl">0.25</span>,<span class="fl">0.333</span>,<span class="fl">0.5</span>,<span class="fl">0.666</span>,<span class="fl">0.75</span>), <span class="dt">col=</span><span class="dv">8</span>)
<span class="kw">points</span>(<span class="kw">getPOS</span>(vcf), freq1[,mySample], <span class="dt">pch =</span> <span class="dv">20</span>, <span class="dt">col=</span> <span class="st">&quot;#A6CEE344&quot;</span>)
<span class="kw">points</span>(<span class="kw">getPOS</span>(vcf), freq2[,mySample], <span class="dt">pch =</span> <span class="dv">20</span>, <span class="dt">col=</span> <span class="st">&quot;#1F78B444&quot;</span>)
<span class="kw">segments</span>(<span class="dt">x0=</span>myPeaks1$wins[,<span class="st">&#39;START_pos&#39;</span>], <span class="dt">y0=</span>myPeaks1$peaks[,mySample],
         <span class="dt">x1=</span>myPeaks1$wins[,<span class="st">&#39;END_pos&#39;</span>], <span class="dt">lwd=</span><span class="dv">3</span>)
<span class="kw">segments</span>(<span class="dt">x0=</span>myPeaks1$wins[,<span class="st">&#39;START_pos&#39;</span>], <span class="dt">y0=</span>myPeaks2$peaks[,mySample],
         <span class="dt">x1=</span>myPeaks1$wins[,<span class="st">&#39;END_pos&#39;</span>], <span class="dt">lwd=</span><span class="dv">3</span>)
}</code></pre></div>
<p><img src="determining_ploidy_2_files/figure-html/unnamed-chunk-9-1.png" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</code></pre></div>
<p>We now have a way to create windows of allele balance ratios. We also have several ways to visualize this data. Perhaps most importantly, we have a numerical summary of the peak in each window so we can loop over all of the chromosomes in our reference.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
