<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Brian J. Knaus" />


<title>Haploidizing diploid data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">vcfR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="rlanguage.html">R language</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="quick_intro.html">A quick introduction</a>
    </li>
    <li>
      <a href="vcf_data.html">VCF data</a>
    </li>
    <li>
      <a href="vcfR_object.html">vcfR objects</a>
    </li>
    <li>
      <a href="matrices.html">Extracting matrices</a>
    </li>
    <li>
      <a href="tidy_vcfR.html">Tidy vcfR</a>
    </li>
    <li>
      <a href="chromR_object.html">chromR objects</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    GBS class
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="gbs_class.html">GBS class</a>
    </li>
    <li>
      <a href="vcf_data.html">VCF data</a>
    </li>
    <li>
      <a href="extract_data.html">Extract data</a>
    </li>
    <li>
      <a href="depth_plot.html">Depth plot</a>
    </li>
    <li>
      <a href="missing_data.html">Missing data</a>
    </li>
    <li>
      <a href="censoring_data.html">Censoring data</a>
    </li>
    <li>
      <a href="omitting_data.html">Omitting data</a>
    </li>
    <li>
      <a href="apply.html">Apply</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Ploidy
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="ploidy.html">Ploidy</a>
    </li>
    <li>
      <a href="determining_ploidy_1.html">Determining ploidy 1</a>
    </li>
    <li>
      <a href="determining_ploidy_2.html">Determining ploidy 2</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Export
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="export.html">Overview</a>
    </li>
    <li>
      <a href="export_vcfgz.html">Export to *vcf.gz</a>
    </li>
    <li>
      <a href="export_genind_genclone.html">Genind and Genclone</a>
    </li>
    <li>
      <a href="export_genlight_snpclone.html">Genlight and SNPclone</a>
    </li>
    <li>
      <a href="dnabin.html">DNAbin</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    FAQ
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="subset_data_to_1chrom.html">Subset to 1 chromosome</a>
    </li>
    <li>
      <a href="vcf_software.html">VCF software</a>
    </li>
    <li>
      <a href="dip_to_hap.html">Haploidizing diploid data</a>
    </li>
    <li>
      <a href="compiling_vcfR.html">Compiling vcfR</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<p>
<center>
<h3>vcfR documentation</h3>
by
<br>
Brian J. Knaus and Niklaus J. Gr&uuml;nwald
</center>
</p>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Haploidizing diploid data</h1>
<h4 class="author"><em>Brian J. Knaus</em></h4>
<h4 class="date"><em>February 7, 2017</em></h4>

</div>


<p>Load example data set and extract the genotypes.</p>
<pre class="r"><code>library(vcfR)
data(vcfR_test)
head(vcfR_test)</code></pre>
<pre><code>## [1] &quot;***** Object of class &#39;vcfR&#39; *****&quot;
## [1] &quot;***** Meta section *****&quot;
## [1] &quot;##fileformat=VCFv4.3&quot;
## [1] &quot;##fileDate=20090805&quot;
## [1] &quot;##source=myImputationProgramV3.1&quot;
## [1] &quot;##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta&quot;
## [1] &quot;##contig=&lt;ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d [Truncated]&quot;
## [1] &quot;##phasing=partial&quot;
## [1] &quot;First 6 rows.&quot;
## [1] 
## [1] &quot;***** Fixed section *****&quot;
##      CHROM POS       ID          REF   ALT      QUAL FILTER
## [1,] &quot;20&quot;  &quot;14370&quot;   &quot;rs6054257&quot; &quot;G&quot;   &quot;A&quot;      &quot;29&quot; &quot;PASS&quot;
## [2,] &quot;20&quot;  &quot;17330&quot;   NA          &quot;T&quot;   &quot;A&quot;      &quot;3&quot;  &quot;q10&quot; 
## [3,] &quot;20&quot;  &quot;1110696&quot; &quot;rs6040355&quot; &quot;A&quot;   &quot;G,T&quot;    &quot;67&quot; &quot;PASS&quot;
## [4,] &quot;20&quot;  &quot;1230237&quot; NA          &quot;T&quot;   NA       &quot;47&quot; &quot;PASS&quot;
## [5,] &quot;20&quot;  &quot;1234567&quot; &quot;microsat1&quot; &quot;GTC&quot; &quot;G,GTCT&quot; &quot;50&quot; &quot;PASS&quot;
## [1] 
## [1] &quot;***** Genotype section *****&quot;
##      FORMAT        NA00001          NA00002          NA00003       
## [1,] &quot;GT:GQ:DP:HQ&quot; &quot;0|0:48:1:51,51&quot; &quot;1|0:48:8:51,51&quot; &quot;1/1:43:5:.,.&quot;
## [2,] &quot;GT:GQ:DP:HQ&quot; &quot;0|0:49:3:58,50&quot; &quot;0|1:3:5:65,3&quot;   &quot;0/0:41:3&quot;    
## [3,] &quot;GT:GQ:DP:HQ&quot; &quot;1|2:21:6:23,27&quot; &quot;2|1:2:0:18,2&quot;   &quot;2/2:35:4&quot;    
## [4,] &quot;GT:GQ:DP:HQ&quot; &quot;0|0:54:7:56,60&quot; &quot;0|0:48:4:51,51&quot; &quot;0/0:61:2&quot;    
## [5,] &quot;GT:GQ:DP&quot;    &quot;0/1:35:4&quot;       &quot;0/2:17:2&quot;       &quot;1/1:40:3&quot;    
## [1] 
## [1] &quot;Unique GT formats:&quot;
## [1] &quot;GT:GQ:DP:HQ&quot; &quot;GT:GQ:DP&quot;   
## [1]</code></pre>
<pre class="r"><code>gt &lt;- extract.gt(vcfR_test)
head(gt)</code></pre>
<pre><code>##            NA00001 NA00002 NA00003
## rs6054257  &quot;0|0&quot;   &quot;1|0&quot;   &quot;1/1&quot;  
## 20_17330   &quot;0|0&quot;   &quot;0|1&quot;   &quot;0/0&quot;  
## rs6040355  &quot;1|2&quot;   &quot;2|1&quot;   &quot;2/2&quot;  
## 20_1230237 &quot;0|0&quot;   &quot;0|0&quot;   &quot;0/0&quot;  
## microsat1  &quot;0/1&quot;   &quot;0/2&quot;   &quot;1/1&quot;</code></pre>
<p>Use <code>is_het()</code> to identify heterozygous positions and censor as NA.</p>
<pre class="r"><code>is.na(vcfR_test@gt[,-1][is_het(gt)]) &lt;- TRUE
vcfR_test</code></pre>
<pre><code>## ***** Object of Class vcfR *****
## 3 samples
## 1 CHROMs
## 5 variants
## Object size: 0 Mb
## 40 percent missing data
## *****        *****         *****</code></pre>
<p>Extract the remaining genotypes and use <code>unique()</code> to query for the unique genotypes.</p>
<pre class="r"><code>gt &lt;- extract.gt(vcfR_test)
unique(as.vector(gt))</code></pre>
<pre><code>## [1] &quot;0|0&quot; NA    &quot;1/1&quot; &quot;0/0&quot; &quot;2/2&quot;</code></pre>
<p>We can now manually set each genotype to a haploid state. Note that your genotypes will be different!</p>
<pre class="r"><code>gt[gt==&quot;0|0&quot;] &lt;- 0
gt[gt==&quot;0/0&quot;] &lt;- 0
gt[gt==&quot;1/1&quot;] &lt;- 1
gt[gt==&quot;2/2&quot;] &lt;- 2
head(gt)</code></pre>
<pre><code>##            NA00001 NA00002 NA00003
## rs6054257  &quot;0&quot;     NA      &quot;1&quot;    
## 20_17330   &quot;0&quot;     NA      &quot;0&quot;    
## rs6040355  NA      NA      &quot;2&quot;    
## 20_1230237 &quot;0&quot;     &quot;0&quot;     &quot;0&quot;    
## microsat1  NA      NA      &quot;1&quot;</code></pre>
<p>Now extract everything but the genotype from the vcfR object, paste our haploid genotypes to it and update the vcfR object.</p>
<pre class="r"><code>gt2 &lt;- extract.gt(vcfR_test, extract = FALSE)
head(gt2)</code></pre>
<pre><code>##            NA00001      NA00002      NA00003   
## rs6054257  &quot;48:1:51,51&quot; NA           &quot;43:5:.,.&quot;
## 20_17330   &quot;49:3:58,50&quot; NA           &quot;41:3&quot;    
## rs6040355  NA           NA           &quot;35:4&quot;    
## 20_1230237 &quot;54:7:56,60&quot; &quot;48:4:51,51&quot; &quot;61:2&quot;    
## microsat1  NA           NA           &quot;40:3&quot;</code></pre>
<pre class="r"><code>gt &lt;- matrix( paste(gt, gt2, sep=&quot;:&quot;), nrow=nrow(gt), dimnames=dimnames(gt) )
is.na(gt[gt == &quot;NA:NA&quot;]) &lt;- TRUE
vcfR_test@gt[,-1] &lt;- gt
head(vcfR_test)</code></pre>
<pre><code>## [1] &quot;***** Object of class &#39;vcfR&#39; *****&quot;
## [1] &quot;***** Meta section *****&quot;
## [1] &quot;##fileformat=VCFv4.3&quot;
## [1] &quot;##fileDate=20090805&quot;
## [1] &quot;##source=myImputationProgramV3.1&quot;
## [1] &quot;##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta&quot;
## [1] &quot;##contig=&lt;ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d [Truncated]&quot;
## [1] &quot;##phasing=partial&quot;
## [1] &quot;First 6 rows.&quot;
## [1] 
## [1] &quot;***** Fixed section *****&quot;
##      CHROM POS       ID          REF   ALT      QUAL FILTER
## [1,] &quot;20&quot;  &quot;14370&quot;   &quot;rs6054257&quot; &quot;G&quot;   &quot;A&quot;      &quot;29&quot; &quot;PASS&quot;
## [2,] &quot;20&quot;  &quot;17330&quot;   NA          &quot;T&quot;   &quot;A&quot;      &quot;3&quot;  &quot;q10&quot; 
## [3,] &quot;20&quot;  &quot;1110696&quot; &quot;rs6040355&quot; &quot;A&quot;   &quot;G,T&quot;    &quot;67&quot; &quot;PASS&quot;
## [4,] &quot;20&quot;  &quot;1230237&quot; NA          &quot;T&quot;   NA       &quot;47&quot; &quot;PASS&quot;
## [5,] &quot;20&quot;  &quot;1234567&quot; &quot;microsat1&quot; &quot;GTC&quot; &quot;G,GTCT&quot; &quot;50&quot; &quot;PASS&quot;
## [1] 
## [1] &quot;***** Genotype section *****&quot;
##      FORMAT        NA00001        NA00002        NA00003     
## [1,] &quot;GT:GQ:DP:HQ&quot; &quot;0:48:1:51,51&quot; NA             &quot;1:43:5:.,.&quot;
## [2,] &quot;GT:GQ:DP:HQ&quot; &quot;0:49:3:58,50&quot; NA             &quot;0:41:3&quot;    
## [3,] &quot;GT:GQ:DP:HQ&quot; NA             NA             &quot;2:35:4&quot;    
## [4,] &quot;GT:GQ:DP:HQ&quot; &quot;0:54:7:56,60&quot; &quot;0:48:4:51,51&quot; &quot;0:61:2&quot;    
## [5,] &quot;GT:GQ:DP&quot;    NA             NA             &quot;1:40:3&quot;    
## [1] 
## [1] &quot;Unique GT formats:&quot;
## [1] &quot;GT:GQ:DP:HQ&quot; &quot;GT:GQ:DP&quot;   
## [1]</code></pre>

<center>
<hr class="style1">
<p>Copyright &copy; 2017 Brian J. Knaus. All rights reserved.</p>
<p>USDA Agricultural Research Service, Horticultural Crops Research Lab.</p>
</center>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
