---
title: "GBS class"
output:
  html_document
---


This material is intended for a one hour class on managing variant call format (VCF) data generated by the  genotyping-by-sequencing (GBS) method.
This class was a part of [GBS_Winter_2017](http://teaching.cgrb.oregonstate.edu/MCB/Rhodes/GBS_Winter_2017/) at Oregon State University.
It is assumed that the audience understands the GBS method (i.e., how the data were generated) and has a familiarity with the R language.


## Outline

* VCF data
* Extracting data matrices from VCF data
* Plotting depth (or other parameters)
* Quantifying data missingness
* Censoring data outside an acceptance threshold
* Omitting data outside an acceptance threshold
* R's apply function


The dataset we will be using consists of samples of *Phytophthora rubi*, *P. syringae* and *P. plurivora* that are all mapped to a *P. rubi* reference.
We'll see that the other species do not map well to the *P. rubi* genome.
This will provide a good example for troubleshooting batch effects in datasets.


