<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Brian J. Knaus" />


<title>How much memory do I need?</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">vcfR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="rlanguage.html">R language</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="quick_intro.html">A quick introduction</a>
    </li>
    <li>
      <a href="vcf_data.html">VCF data</a>
    </li>
    <li>
      <a href="vcfR_object.html">vcfR objects</a>
    </li>
    <li>
      <a href="how_much_memory.html">How much memory</a>
    </li>
    <li>
      <a href="matrices.html">Extracting matrices</a>
    </li>
    <li>
      <a href="tidy_vcfR.html">Tidy vcfR</a>
    </li>
    <li>
      <a href="chromR_object.html">chromR objects</a>
    </li>
    <li>
      <a href="genetic_differentiation.html">Genetic differentiation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    GBS class
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="gbs_class.html">GBS class</a>
    </li>
    <li>
      <a href="vcf_data.html">VCF data</a>
    </li>
    <li>
      <a href="extract_data.html">Extract data</a>
    </li>
    <li>
      <a href="depth_plot.html">Depth plot</a>
    </li>
    <li>
      <a href="missing_data.html">Missing data</a>
    </li>
    <li>
      <a href="censoring_data.html">Censoring data</a>
    </li>
    <li>
      <a href="omitting_data.html">Omitting data</a>
    </li>
    <li>
      <a href="apply.html">Apply</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Ploidy
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="ploidy.html">Ploidy</a>
    </li>
    <li>
      <a href="determining_ploidy_1.html">Determining ploidy 1</a>
    </li>
    <li>
      <a href="determining_ploidy_2.html">Determining ploidy 2</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Export
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="export.html">Overview</a>
    </li>
    <li>
      <a href="export_vcfgz.html">Export to *vcf.gz</a>
    </li>
    <li>
      <a href="export_genind_genclone.html">Genind and Genclone</a>
    </li>
    <li>
      <a href="export_genlight_snpclone.html">Genlight and SNPclone</a>
    </li>
    <li>
      <a href="dnabin.html">DNAbin</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    FAQ
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="subset_data_to_1chrom.html">Subset to 1 chromosome</a>
    </li>
    <li>
      <a href="vcf_software.html">VCF software</a>
    </li>
    <li>
      <a href="dip_to_hap.html">Haploidizing diploid data</a>
    </li>
    <li>
      <a href="compiling_vcfR.html">Compiling vcfR</a>
    </li>
    <li>
      <a href="reporting_issue.html">Reporting an issue</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<p>
<center>
<h3>vcfR documentation</h3>
by
<br>
Brian J. Knaus and Niklaus J. Gr&uuml;nwald
</center>
</p>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">How much memory do I need?</h1>
<h4 class="author"><em>Brian J. Knaus</em></h4>
<h4 class="date"><em>June 7, 2017</em></h4>

</div>


<p>The R package vcfR attempts to read all data into memory (RAM) in order to perform manipulations or analyses. This is similar to other R packages. The catch with vcfR is that VCF files frequently contain large quantities of data. This may create a situation where you have more data than you can read into memory. Here I explore the memory footprint of a vcfR object to help you decide whether you have enough memory to read in your entire file, or whether you need to develop strategies for working on subsets of this data.</p>
<div id="the-vcfr-object" class="section level2">
<h2>The vcfR object</h2>
<p>The vcfR object is an S4 class object consisting of three slots. The <code>meta</code> slot is a character vector. It is recommended, but not required, that the meta data include information for each chromosome in the reference. In mature projects, with a small number of chromosomes, the size of this data may be fairly small. In less mature projects, where the reference may consist of thousands of contigs, this region may become noticable in size. However, it is likely to be small relative to the rest of the VCF data. The <code>fix</code> slot is a character matrix that always consists of eight columns but has as many rows as there are variants in the VCF file, plus one more row for the header. The <code>gt</code> slot is a character matrix that contains a column for each sample, plus one column to designate the format for each variant, and it has as one row for each variant. As the number of variants in your data grow the sizes of the <code>fix</code> and <code>gt</code> slots grow to accomodate these variants. As the number of samples in your dataset grow the number of columns in your <code>gt</code> slot grows. Here we’ll explore how many variants may require how much memory.</p>
</div>
<div id="simulation-of-data" class="section level2">
<h2>Simulation of data</h2>
<p>In order to estimate the memory footprint of vcfR objects containing different numbers of variants we’ll use the R package <code>memuse</code>. For simplicity, we’ll ignore the <code>meta</code> slot and instead of using both a <code>fix</code> and a <code>gt</code> slot we’ll use a single matrix of 10 columns. This approximates VCF data containing one sample. As your sample size increases, so will your memory footprint. Depending on your comfort level with R, you may be able to modify the below code to match your data set.</p>
<pre class="r"><code>library(&#39;memuse&#39;)</code></pre>
<pre><code>## 
## Attaching package: &#39;memuse&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:ggplot2&#39;:
## 
##     unit</code></pre>
<pre class="r"><code>nvar &lt;- 10^{2:8}
nMb &lt;- howbig(nrow=10, ncol=10, unit=&quot;MB&quot;)

for(i in nvar){
  nMb &lt;- c(nMb, howbig(nrow=i, ncol=8, unit=&quot;KB&quot;))
}
nvar &lt;- c(10, nvar)

nMb &lt;- as.numeric(lapply(nMb, as.numeric))/1000</code></pre>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>Once we’ve simulated our data, we can visualize it. Here we’ll use a simple line graph.</p>
<pre class="r"><code>par(mar=c(5,5,4,2))
plot(log10(nvar), log10(nMb), xaxt=&quot;n&quot;, yaxt=&quot;n&quot;, type=&#39;b&#39;, xlab=&quot;Number of variants&quot;, ylab = &quot;&quot;)
axis(side = 1, at = 1:8, labels=c(&#39;10&#39;, &#39;100&#39;, &#39;1k&#39;, &#39;10k&#39;, &#39;100k&#39;, &#39;1m&#39;, &#39;10m&#39;, &#39;100m&#39;))
axis(side = 2, at = 1:8, labels=c(&#39;800k&#39;, &#39;6.4Mb&#39;, &#39;64Mb&#39;, &#39;640Mb&#39;, &#39;6.4Gb&#39;, &#39;64Gb&#39;, &#39;640Gb&#39;, &#39;6.4Tb&#39;), las=2)
title(ylab=&quot;Memory use&quot;, line=4)
abline(h=1:8, lwd=2, col=&quot;#C0C0C066&quot;)
abline(v=1:8, lwd=2, col=&quot;#C0C0C066&quot;)</code></pre>
<p><img src="how_much_memory_files/figure-html/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mar=c(5,4,4,2))</code></pre>
<p>This should give us an idea of how much data we can read into a vcfR object. When we have 100,000 variants we’ll need about 640 Mb of memory. If we have 1 million variants we’ll need about 6.4 Gb of memory. This is asking a lot of your typical work station. In the past I’ve found that creating objects in R that are over about 1 GB starts to affect the performance of the system. This may change in the future as R is constantly improving.</p>
</div>
<div id="dont-know-how-many-variants-are-in-your-file" class="section level2">
<h2>Don’t know how many variants are in your file?</h2>
<p>If you don’t know how many variants are in your VCF file you may be a t aloss for guessing how much memory you need. If you’re working on a flavor of Unix (OSX, Linux, etc.) you can use the shell to get an idea.</p>
<pre class="bash"><code>grep -v &quot;^#&quot; myVcfFile.vcf | head -n 1000000 | wc -l</code></pre>
<p>Or for gzipped files.</p>
<pre class="bash"><code>zgrep -v &quot;^#&quot; myVcfFile.vcf.gz | head -n 1000000 | wc -l</code></pre>
<p>The <code>zgrep -v &quot;^#&quot; myVcfFile.vcf.gz</code> command sends the contents of your VCF file to standard output, while omitting the meta and header information which have lines begining with a pound or hash sign. By piping this to head you limit the number of lines you count. If you have a really large number of variants you may not want to count them all. By piping this to <code>wc -l</code> it should count the lines for you where each variant is one line. This should help you get an idea of how many variants are in your file.</p>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>Our lab works on plant pathogens that have genomes in the range of tens to hundreds of millions of base pairs in size. We typically work on complete GBS datasets consisting of over 100 samples. These datasets appear to read into memory and work quite well. For genomic projects we typically have more variants than we can comfortably read into memory. For these projects we typically work on a per supercontig manner as a way of decomposing a genomic project into more manageable subsets. The function <code>read.vcfR()</code> includes parameters <code>nrows</code> and <code>skip</code> to allow you to read portions of VCF files into memory. Note that this will require serial access so that selecting regions of a file that are near the end of the file may require some time to access. I find that breaking a VCF file into several files, one for each supercontig, helps me sort through a genome more efficiently. Hopefully this information will help you as well.</p>
</div>

<center>
<hr class="style1">
<p>Copyright &copy; 2017 Brian J. Knaus. All rights reserved.</p>
<p>USDA Agricultural Research Service, Horticultural Crops Research Lab.</p>
</center>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
