---
title: "Missing data"
output:
  html_document:
    toc: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = 'center')
knitr::opts_chunk$set(fig.width = 12)
```



As the size of our dataset grow in terms of samples and variants the size of our data matrix grows.
As the size of our data matrix grows it also increases the opportunity to have missing data.
Also, some of our quality filtering steps increased the degree of missingness in our data matrix by setting values that we determined to be of unusual quality to NA.
One way of managing missing data is to use interpolation, a set of methods that attempts to infer what the most likely genotype should be and replaces the missing genotype with the interpolated genotype.
However, if your data has a large degree of missingness you may want to attempt to mitigate missingness instead of interpolation or prior to an interpolation step.
Missing data can frequently be due to samples (columns) or variants (rows) of low quality.
Here we demonstrate how to identify samples and variants in the data set that have a high degree of missingness and omit them.


```{r, results='hide', echo = FALSE}
suppressPackageStartupMessages( library(vcfR) )
vcf <- read.vcfR('TASSEL_GBS0077.vcf.gz')
dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)
```


## Quantifying missingness, one sample


To quantify missingness for a single sample we can use the fnction `is.na()`.
This function uses a vector as an argument and returns a logical vector (TRUE and FALSE) indicating which values are NA.
If we remind ourselves that TRUEs and FALSEs are numerically encoded as ones and zeros it reminds us we can take a sum of this logical vector to determine the degree of missingness.


```{r}
as.numeric(TRUE)
sum(is.na(dp[,1]))
# apply(dp, MARGIN = 2, function(x){sum(is.na(x))})
```


This reports the number of missing variants in the first sample.
We could similarly count the number of missing samples from a variant by accessing a row instead of a column.
We could also convert this to a percentage by using `length()` to determine the total number of values in either the column or row and use this as a denominator.


## Quantifying missingness, all samples


This is illustrative of what we can accomplish for a single sample or variant.
We typically have many samples an tens of thousand (or more) variants.
We can extend the functionality of the above example to many columns or rows by using the `apply()` function.
See the section on apply if you are unfamiliar with this function.
Because we will be summarizing many samples we will use a barplot to visualize the results as opposed to trying to scrutinize the numerical information.


```{r}
myMiss <- apply(dp, MARGIN = 2, function(x){ sum(is.na(x)) })
myMiss <- myMiss/nrow(vcf)

library(RColorBrewer)
palette(brewer.pal(n=12, name = 'Set3'))

par(mar = c(12,4,4,2))
barplot(myMiss, las = 2, col = 1:12)
title(ylab = "Missingness (%)")
par(mar = c(5,4,4,2))
```


We see that there are two classes of samples.
The samples with long names have a high degree of missingness (around 80%) while the samples with short names have a relatively low degree of missingness.


We can do something similar to query the variants (rows) for missingness.
However, our dataset has over 60 thousand variants, so we can't visualize this with a barchart with 60 thousand bars.
Instead of usung a barchart we'll use a histogram.



```{r, fig.width=5}
myMiss <- apply(dp, MARGIN = 1, function(x){ sum(is.na(x)) })
myMiss <- myMiss/ncol(vcf@gt[,-1])

hist(myMiss, col = "#8DD3C7", xlab = "Missingness (%)")
```
