<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Quality filtering</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">vcfR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="rlanguage.html">R language</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="quick_intro.html">A quick introduction</a>
    </li>
    <li>
      <a href="vcf_data.html">VCF data</a>
    </li>
    <li>
      <a href="vcfR_object.html">vcfR objects</a>
    </li>
    <li>
      <a href="matrices.html">Extracting matrices</a>
    </li>
    <li>
      <a href="tidy_vcfR.html">Tidy vcfR</a>
    </li>
    <li>
      <a href="chromR_object.html">chromR objects</a>
    </li>
    <li>
      <a href="genetic_differentiation.html">Genetic differentiation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    GBS class
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="gbs_class.html">GBS class</a>
    </li>
    <li>
      <a href="vcf_data.html">VCF data</a>
    </li>
    <li>
      <a href="extract_data.html">Extract data</a>
    </li>
    <li>
      <a href="depth_plot.html">Depth plot</a>
    </li>
    <li>
      <a href="missing_data.html">Missing data</a>
    </li>
    <li>
      <a href="censoring_data.html">Censoring data</a>
    </li>
    <li>
      <a href="omitting_data.html">Omitting data</a>
    </li>
    <li>
      <a href="apply.html">Apply</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Ploidy
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="ploidy.html">Ploidy</a>
    </li>
    <li>
      <a href="determining_ploidy_1.html">Determining ploidy 1</a>
    </li>
    <li>
      <a href="determining_ploidy_2.html">Determining ploidy 2</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Export
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="export.html">Overview</a>
    </li>
    <li>
      <a href="export_vcfgz.html">Export to *vcf.gz</a>
    </li>
    <li>
      <a href="export_genind_genclone.html">Genind and Genclone</a>
    </li>
    <li>
      <a href="export_genlight_snpclone.html">Genlight and SNPclone</a>
    </li>
    <li>
      <a href="dnabin.html">DNAbin</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    FAQ
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="subset_data_to_1chrom.html">Subset to 1 chromosome</a>
    </li>
    <li>
      <a href="vcf_software.html">VCF software</a>
    </li>
    <li>
      <a href="dip_to_hap.html">Haploidizing diploid data</a>
    </li>
    <li>
      <a href="compiling_vcfR.html">Compiling vcfR</a>
    </li>
    <li>
      <a href="reporting_issue.html">Reporting an issue</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<p>
<center>
<h3>vcfR documentation</h3>
by
<br>
Brian J. Knaus and Niklaus J. Gr&uuml;nwald
</center>
</p>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Quality filtering</h1>

</div>


<p>Once a sequencing run has been completed the sequence reads are typically aligned to a reference genome and variants are called. These variants are typically stored in a VCF format file. THe R package vcfR was designed to import and export this data into the R environment. Once in the R environment, vcfR provides tools to manipulate the data to help you understand its content. Common issues in variant data may be that certain samples may be of low quality or certain vriants may be of questionable quality. Here we’ll provide examples of how vcfR can help accomplish these tasks.</p>
<div id="data-import" class="section level2">
<h2>Data import</h2>
<p>Our first step will be to import the VCF data into R. The result will be that our VCF data will be stored in a ‘vcfR’ object. Once we have read it in we can validate our data is what we expect by using the <code>show</code> method. The <code>show</code> method is available for many types of objects and is implemented when the object’s name is entered at the console with no other information. THe <code>show</code> method typically provides summary information about the object and what it contains.</p>
<pre class="r"><code>library(vcfR)
vcf &lt;- read.vcfR(&#39;TASSEL_GBS0077.vcf.gz&#39;)
class(vcf)</code></pre>
<pre class="r"><code>vcf</code></pre>
<pre><code>## ***** Object of Class vcfR *****
## 61 samples
## 7171 CHROMs
## 69,296 variants
## Object size: 46.5 Mb
## 37.62 percent missing data
## *****        *****         *****</code></pre>
<p>The results from the <code>show</code> method informs us that our vcfR object contains 61 samples and 69,296 variants. This is what we expected so we can be confident that our data was imported correctly. The output also informs us that there is just over 37% missing data in our file. In our experience this is typical for GBS data sets.</p>
</div>
<div id="extract-depth-dp" class="section level2">
<h2>Extract depth (DP)</h2>
<p>The vcfR function <code>extract.gt()</code> is used to extract matrices of data from teh GT portion of VCF data. the funtion <code>extract.gt()</code> provides a link between VCF data and R. Much of R is designed to operate on matrices of data and once <code>extract.gt()</code> provides this matrix the universe of R becomes available. Note that we use the ‘as.numeric=TRUE’ option here. We should only use this option when we are certain that we have numeric data. If you use it on non-numeric data R will do its best to do something, which is not likely to be what you expect. We can use the <code>queryMETA()</code> function remind us what this element is.</p>
<pre class="r"><code>queryMETA(vcf, element = &#39;FORMAT.+DP&#39;)</code></pre>
<pre><code>## [[1]]
## [1] &quot;FORMAT=ID=DP&quot;                                                 
## [2] &quot;Number=1&quot;                                                     
## [3] &quot;Type=Integer&quot;                                                 
## [4] &quot;Description=Read Depth (only filtered reads used for calling)&quot;</code></pre>
<pre class="r"><code>vcf@gt[1:4,1:4]</code></pre>
<pre><code>##      FORMAT           1                       10                       
## [1,] &quot;GT:AD:DP:GQ:PL&quot; &quot;0/0:5,0:5:96:0,15,180&quot; &quot;1/1:0,10:10:99:255,30,0&quot;
## [2,] &quot;GT:AD:DP:GQ:PL&quot; &quot;1/1:0,1:1:66:36,3,0&quot;   &quot;0/0:4,0:4:94:0,12,144&quot;  
## [3,] &quot;GT:AD:DP:GQ:PL&quot; NA                      NA                       
## [4,] &quot;GT:AD:DP:GQ:PL&quot; &quot;1/1:0,5:5:96:180,15,0&quot; &quot;0/0:8,0:8:99:0,24,255&quot;  
##      13                     
## [1,] &quot;0/0:9,0:9:99:0,27,255&quot;
## [2,] &quot;0/0:1,0:1:66:0,3,36&quot;  
## [3,] &quot;0/0:3,0:3:88:0,9,108&quot; 
## [4,] NA</code></pre>
<pre class="r"><code>dp &lt;- extract.gt(vcf, element = &quot;DP&quot;, as.numeric=TRUE)
dp[1:4,1:3]</code></pre>
<pre><code>##          1 10 13
## S1_4509  5 10  9
## S1_4657  1  4  1
## S1_5193 NA NA  3
## S1_5647  5  8 NA</code></pre>
</div>
<div id="missing-data" class="section level2">
<h2>Missing data</h2>
<pre class="r"><code>sum(is.na(dp[,1]))</code></pre>
<pre><code>## [1] 11725</code></pre>
<pre class="r"><code># apply(dp, MARGIN = 2, function(x){sum(is.na(x))})</code></pre>
<pre class="r"><code>myMiss &lt;- apply(dp, MARGIN = 2, function(x){sum(is.na(x))})
myMiss &lt;- myMiss/nrow(vcf)

library(RColorBrewer)
palette(brewer.pal(n=12, name = &#39;Set3&#39;))

par(mar = c(12,4,4,2))
barplot(myMiss, las = 2, col = 1:12)
title(ylab = &quot;Missingness&quot;)</code></pre>
<p><img src="quality_filtering_files/figure-html/unnamed-chunk-5-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mar = c(5,4,4,2))</code></pre>
</div>
<div id="sequence-depth" class="section level2">
<h2>Sequence depth</h2>
<pre class="r"><code>boxplot(dp, col=2:8, las=3)</code></pre>
<p><img src="quality_filtering_files/figure-html/unnamed-chunk-6-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>library(reshape2)
library(ggplot2) 
library(cowplot)

# Melt our matrix into a long form data.frame.
dpf &lt;- melt(dp, varnames=c(&#39;Index&#39;, &#39;Sample&#39;), value.name = &#39;Depth&#39;, na.rm=TRUE)
dpf &lt;- dpf[ dpf$Depth &gt; 0,]

# Create a row designator.
# You may want to adjust this
#samps_per_row &lt;- 20
samps_per_row &lt;- 16
myRows &lt;- ceiling(length(levels(dpf$Sample))/samps_per_row)
myList &lt;- vector(mode = &quot;list&quot;, length = myRows)

for(i in 1:myRows){
  myIndex &lt;- c(i*samps_per_row - samps_per_row + 1):c(i*samps_per_row)
  myIndex &lt;- myIndex[myIndex &lt;= length(levels(dpf$Sample))]
  myLevels &lt;- levels(dpf$Sample)[myIndex]
  myRegex &lt;- paste(myLevels, collapse = &quot;$|^&quot;)
  myRegex &lt;- paste(&quot;^&quot;, myRegex, &quot;$&quot;, sep = &quot;&quot;)
  myList[[i]] &lt;- dpf[grep(myRegex, dpf$Sample),]
  myList[[i]]$Sample &lt;- factor(myList[[i]]$Sample)
}

# Create the plot.
myPlots &lt;- vector(mode = &quot;list&quot;, length = myRows)
for(i in 1:myRows){
  myPlots[[i]] &lt;- ggplot(myList[[i]], aes(x=Sample, y=Depth)) + 
                  geom_violin(fill=&quot;#90EE90&quot;, adjust=1.0, scale = &quot;count&quot;, trim=TRUE)

  myPlots[[i]] &lt;- myPlots[[i]] + theme_bw()
  myPlots[[i]] &lt;- myPlots[[i]] + theme(axis.title.x = element_blank(), 
                  axis.text.x = element_text(angle = 60, hjust = 1))
  myPlots[[i]] &lt;- myPlots[[i]] + scale_y_continuous(trans=scales::log2_trans(), 
                  breaks=c(1, 10, 100, 800),
                  minor_breaks=c(1:10, 2:10*10, 2:8*100))
  myPlots[[i]] &lt;- myPlots[[i]] + theme( panel.grid.major.y=element_line(color = &quot;#A9A9A9&quot;, size=0.6) )
  myPlots[[i]] &lt;- myPlots[[i]] + theme( panel.grid.minor.y=element_line(color = &quot;#C0C0C0&quot;, size=0.2) )
}</code></pre>
<pre class="r"><code># Plot the plot.
plot_grid(plotlist = myPlots, nrow = myRows)</code></pre>
<p><img src="quality_filtering_files/figure-html/unnamed-chunk-8-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>Once we have extracted the quantile information we can use apply and quantile to build intervals for what we may consider acceptable coverage.</p>
<pre class="r"><code>quants &lt;- apply(dp, MARGIN=2, quantile, probs=c(0.1, 0.8), na.rm=TRUE)
#quants &lt;- apply(dp, MARGIN=2, quantile, probs=c(0.34, 0.68), na.rm=TRUE)
quants[,1:17]</code></pre>
<pre><code>##      1 10 13 1725-d-8-21-07-r-1-pm-r 18 1858-d-10-15-07-p-3-pm-r  2 20 22
## 10%  1  2  1                       1  1                        1  2  2  2
## 80% 18 42 17                      15 25                       16 34 31 37
##     31 34 3450-d-4-06-09-r-3-pm-r 37 39  4 40 43
## 10%  2  1                       1  2  2  2  2  1
## 80% 29 26                      12 27 36 27 33 23</code></pre>
<p>We can now use these thresholds to censor data outside this threshold.</p>
<pre class="r"><code>dp2 &lt;- sweep(dp, MARGIN=2, FUN = &quot;-&quot;, quants[1,])
dp[dp2 &lt; 0] &lt;- NA

dp2 &lt;- sweep(dp, MARGIN=2, FUN = &quot;-&quot;, quants[2,])
dp[dp2 &gt; 0] &lt;- NA

dp[dp &lt; 4] &lt;- NA</code></pre>
<p>Update the vcfR object with the censored data.</p>
<pre class="r"><code>vcf@gt[,-1][ is.na(dp) == TRUE ] &lt;- NA</code></pre>
<p>We’ll want to see how this has affected the missingness of our vcfR object.</p>
<pre class="r"><code>vcf</code></pre>
<pre><code>## ***** Object of Class vcfR *****
## 61 samples
## 7171 CHROMs
## 69,296 variants
## Object size: 45.2 Mb
## 66.63 percent missing data
## *****        *****         *****</code></pre>
<p>We’ll want to mitigate variants with a high degree of missingness.</p>
<pre class="r"><code>dp &lt;- extract.gt(vcf, element = &quot;DP&quot;, as.numeric=TRUE)

miss &lt;- apply(dp, MARGIN=1, function(x){sum(is.na(x))})
miss &lt;- miss/ncol(dp)</code></pre>
<p>Plot a histogram.</p>
<pre class="r"><code>hist(miss, col=8, breaks=seq(0,1,by=0.02))</code></pre>
<p><img src="quality_filtering_files/figure-html/unnamed-chunk-14-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>Omit variants with a high degree of missingness.</p>
<pre class="r"><code>#vcf &lt;- vcf[miss &lt; 0.05,]
vcf &lt;- vcf[miss &lt; 0.1,]
vcf</code></pre>
<pre><code>## ***** Object of Class vcfR *****
## 61 samples
## 102 CHROMs
## 295 variants
## Object size: 0.2 Mb
## 7.719 percent missing data
## *****        *****         *****</code></pre>
<pre class="r"><code>dp &lt;- extract.gt(vcf, element = &quot;DP&quot;, as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)</code></pre>
<p><img src="quality_filtering_files/figure-html/unnamed-chunk-16-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>#heatmap.bp(dp[1:1000,], rlabels = FALSE)</code></pre>
<pre class="r"><code>boxplot(dp, col=2:8, las=3)</code></pre>
<p><img src="quality_filtering_files/figure-html/unnamed-chunk-17-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
<div id="output-to-file" class="section level2">
<h2>Output to file</h2>
<pre class="r"><code>write.vcf(tas, &#39;TASSEL_GBS0077_dp_filtered.vcf.gz&#39;)</code></pre>
</div>

<center>
<hr class="style1">
<p>Copyright &copy; 2017 Brian J. Knaus. All rights reserved.</p>
<p>USDA Agricultural Research Service, Horticultural Crops Research Lab.</p>
</center>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
